src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceSetDeserializer.cs

```diff
@@ -147,10 +192,40 @@ public virtual IEnumerable ReadResourceSet(ODataResourceSetWrapper resourceSet,
                     Error.Format(SRResources.TypeCannotBeDeserialized, elementType.FullName()));
             }
 
-            foreach (ODataResourceWrapper resourceWrapper in resourceSet.Resources)
+            //Ideally we don't need to special case ResourceSetType.ResourceSet, since the code that handles a deltaresourcesetwrapper will also handle a ResourceSetWrapper, 
+            //but it may be more efficient for the common case.
+
+            if (resourceSet.ResourceSetType == ResourceSetType.ResourceSet)
             {
-                yield return deserializer.ReadInline(resourceWrapper, elementType, readContext);
+                foreach (ODataResourceWrapper resourceWrapper in resourceSet.Resources)
+                {
+                    yield return deserializer.ReadInline(resourceWrapper, elementType, readContext);
+                }
+            }
+            else
+            {
+                Type clrType = EdmLibHelpers.GetClrType(elementType, readContext.Model);                           
+
+                foreach (ODataItemBase odataItemBase in resourceSet.Resources)
```

Comment:
>ODataItemBase [](start = 25, length = 13)

the type of resoureSet.Resources is ODataResourceWrapper, why don't you use ODataResourceWrapper in foreach.

Then we can get rid of line 211